### ISSUE TITLE
Set membership operator for type constraints

### ISSUE THREAD

**Issue by mswrenn** [GaloisInc/cryptol#697]
If an example parameter `p` can be 2, 3, or 6, I can define a type constraint loosely with:
```
type constraint (6 >= p, p >= 3, p >= 2)
``` 
or accurately with:
```
type constraint (6 >= p, p != 5, p != 4, p >= 3, p >= 2)
```
However, once the range of possible values gets even slightly larger, defining the correct constraint quickly gets out of hand. For example if `w` was one of 8, 16, 32, or 64, I would need to write out a lot of inequalities.

Is it possible to add a shorthand for this? Something like:
```
type constraint (p in {2, 3, 6})
type constraint (w in {8, 16, 32, 64})
```

**Comment by brianhuffman** [comment]
Minor quibble: The constraint `(6 >= p, p >= 3, p >= 2)` would actually exclude `p = 2` as a solution, so you'd probably want to leave out `p >= 3`.

But anyway, I agree that this is something that users have expressed interest in before, so we should think seriously about adding a special constraint form for set membership.

I believe that @yav wanted to specifically avoid having arbitrary logical disjunction in our constraint language, because (I believe) it could easily lead to exponential blowup when solving constraints, or something like that. Perhaps a more restricted form of disjunction, like membership in a list of specific numbers, wouldn't have that problem.

**Comment by yav** [comment]
What are some algorithms where such a constraint would be useful?

**Comment by brianhuffman** [comment]
Lots of crypto specifications say that a parameter has to be one of a finite list of possible values. For example, FIPS-197 says that the AES key length can be 128, 192, or 256 bits. It would be nice to be able to express these constraints directly in a Cryptol spec.

**Comment by yav** [comment]
All the examples I've seen are like this, where the constraints tent to have some structure, which
matches how the algorithm underneath works.  For example in this case, it is `i * 64` for `2 <= i <= 4`.

Having an arbitrary set of sizes seems unlikely to come out of an actual algorithm, although, I guess, I could see it possibly being useful in some sort of wrapper function, that hides away some of this structure.

**Comment by weaversa** [comment]
@yav ECDSA key sizes don't fit an arithmetic progression.

@brianhuffman I think the best way to deal with what you suggest is to use a parameterized module, such as we've already done with AES, SHA, and ECDSA.

Maybe Max will say a bit more about his usecase.

**Comment by mswrenn** [comment]
Many of the Round 2 submissions have parameters without this structure as well, e.g. a set of large preselected primes.

And, one note on the parameterized modules: Say I write a parameterized module `M` for my first example of `p` in `{2, 3, 6}`, but opt to use the partial constraint `(6 >= p, p >= 2)` because writing out the correct one is impractical. Then, someone (me) might import it with ```import `M``` and accidentally use a function `f` from `M` with ```f `{p = 4}```, which can lead to many misleading errors. With the full constraint, `4 != 4` is caught immediately.

**Comment by yav** [comment]
@mswrenn do you have a small example of a Cryptol function that works for `{2,3,6}` but not 4?   

Or is the issue that the function would work just fine for 4 (in the sense that it won't crash), but the result would be wrong due to some other reasoning not captured in the types, and that's what you are trying to express?

**Comment by mswrenn** [comment]
@yav right, poor use of the word error on my part. I'm trying to express the latter. Thinking a bit more, I could write a guard manually as a combination of ``` `p```, `if-else`, and `error` in a top level function and export that, but it would be nice to define the full constraint and have it prevent any misuse instead.
